h2o-genmodel/src/main/java/hex/genmodel/algos/tree/ContributionComposer.java

```diff
@@ -0,0 +1,66 @@
+package hex.genmodel.algos.tree;
+
+import hex.genmodel.attributes.parameters.KeyValue;
+import hex.genmodel.utils.ArrayUtils;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.Comparator;
+
+public class ContributionComposer implements Serializable {
+    
+    /**
+     * Sort shapley values and compose desired output
+     *
+     * @param contribs Raw contributions to be composed
+     * @param contribNames Contribution corresponding feature names
+     * @param topN Return only #topN highest contributions + bias.
+     * @param topBottomN Return only #topBottomN lowest contributions + bias
+     *                   If topN and topBottomN are defined together then return array of #topN + #topBottomN + bias
+     * @param abs True to compare absolute values of contributions
+     * @return Sorted KeyValue array of contributions of size #topN + #topBottomN + bias
+     *         If topN < 0 || topBottomN < 0 then all descending sorted contributions is returned.
+     */
+    public final KeyValue[] composeContributions(float[] contribs, String[] contribNames, int topN, int topBottomN, boolean abs) {
+        if (topBottomN == 0) {
+            return composeSortedContributions(contribs, contribNames, topN, new KeyValue.DescComparator(abs));
+        } else if (topN == 0){
+            return composeSortedContributions(contribs, contribNames, topBottomN, new KeyValue.AscComparator(abs));
+        } else if ((topN + topBottomN) >= contribs.length || topN < 0 || topBottomN < 0) {
+            return composeSortedContributions(contribs, contribNames, contribs.length, new KeyValue.DescComparator(abs));
+        }
+        
+        KeyValue[] topSorted = composeSortedContributions(contribs,contribNames, topN, new KeyValue.DescComparator(abs));
```

Comment:
can we sort everything just once? and then pick the result?

also it seems like you are operating with n*logn complexity but the problem only needs n*logk (where k is the "topN"); probably not a big deal since n won't be crazy big

Reply:
> Can we sort everything just once? and then pick the result?

Yes, I fixed the double sorting.

Reply:
> also it seems like you are operating with nlogn complexity but the problem only needs nlogk (where k is the "topN"); probably not a big deal since n won't be crazy big

In a nutshell, I agree that we can be more clever here but IMHO it is enough.

I am operating with nlogn. I found that best known algorithm has linear time. Algorithm with Heap could be nlogk, but AFAIK you have to run it twice when you want to have #topN and #topBottomN. I rather choose a KISS principle and use java build-in sorting and provide a clear interface for sorting functionality in case of future replacement of this part of algorithm.